### PR TITLE
[Substrait to Velox] Add tests to CircleCI

### PR DESCRIPTION
**Pull request by zhejiangxiaomai**
* Upgrade protobuf version to 3.21.4
* Remove usage of deprecated Substrait APIs. Update velox/substrait/tests/data/*.json files to use the latest Substrait format.
* Rename "linux-parquet-hive-storage-adapters-build" to "linux-adapters".
* Add install_protobuf to setup-adapters.sh. 
* Add Substrait tests to CircleCI job **linux-adapters.**

### CODE REVIEW

**Comment by zhejiangxiaomai** [comment]
cc: @rui-mo  @PenghuiJiao

File: scripts/setup-ubuntu.sh
Comment (mbasmanova) [inline scripts/setup-ubuntu.sh:28]: naming: PROTOBUF_VERSION

File: scripts/setup-ubuntu.sh
Comment (mbasmanova) [inline scripts/setup-ubuntu.sh:100]: Should we update other setup-xxx.sh scripts?

@kgpai @majetideepak Krishna, Deepak, would you help review the setup script changes?

File: velox/substrait/SubstraitToVeloxPlan.cpp
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxPlan.cpp:606]: naming: param -> arg and paraExpr -> argExpr

File: velox/substrait/VeloxToSubstraitExpr.cpp
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:145]: This change seems unrelated. It would be nice to move it into a separate PR. 

Also, use VELOX_CHECK and fix the error message:

`VELOX_CHECK(functionMap_.find(functionName) != functionMap_.end(), "Scalar function not found: {}", functionName)`

File: velox/substrait/TypeUtils.h
Comment (mbasmanova) [inline velox/substrait/TypeUtils.h:26]: This function doesn't have anything to do with types, hence, it doesn't belong to a file named TypeUtils.h.

I wonder if we can just remove this function and change the callers to use arg.value() directly.

**Review by mbasmanova — COMMENTED** [review]
@zhejiangxiaomai Thank you for upgrading the protobuf files. Some comments below.

It appears that CircleCI doesn't run Substrait tests, hence, can't tell if these are passing. It would be nice to add these tests to CircleCI runs in a separate PR.

File: scripts/setup-ubuntu.sh
Comment (zhejiangxiaomai) [reply scripts/setup-ubuntu.sh:100]: > Should we update other setup-xxx.sh scripts?
> 
> @kgpai @majetideepak Krishna, Deepak, would you help review the setup script changes?

And please also check if we need do some changes in setup-circleci.sh. because i find "dnf_install protobuf-devel" there,

File: .circleci/config.yml
Comment (rui-mo) [inline .circleci/config.yml:401]: Looks like this action will generate test executables into the `_build/debug` folder. But at below, the tests in release folder are triggered with `cd _build/release && ctest -j 16 -VV --output-on-failure`.

Comment (zhejiangxiaomai) [reply .circleci/config.yml]: > Looks like this action will generate test executables into the `_build/debug` folder. But at below, the tests in release folder are triggered with `cd _build/release && ctest -j 16 -VV --output-on-failure`.

You are correct. Thank you for your review.

Comment (kgpai) [reply .circleci/config.yml:401]: If I am not mistaken tests can be generated for make release also..

File: CMakeLists.txt
Comment (kgpai) [inline CMakeLists.txt:302]: We need to ensure that 3.21.0 works well centos 7  , else this is a no go. Can we validate that 3.21 is available for centos 7 (ideally without having to build from source).

File: .circleci/config.yml
Comment (zhejiangxiaomai) [reply .circleci/config.yml]: > If I am not mistaken tests can be generated for make release also..

I changed "debug" -> "release" now.

File: CMakeLists.txt
Comment (zhejiangxiaomai) [reply CMakeLists.txt]: > We need to ensure that 3.21.0 works well centos 7 , else this is a no go. Can we validate that 3.21 is available for centos 7 (ideally without having to build from source).

I changed "3.21.0" -> "3.0.0" now.

**Comment by zhejiangxiaomai** [comment]
Hi, @kgpai ,I met a problem in Circleci. This PR need new API in higher protobuf version to compile code. But Circle docker environment will not run setup-ubuntu.sh or setup-circleci.sh to upgrade protobuf. So I can not pass compile now. Could you please give me some suggestions about it?   I originally wanted to add setup-ubuntu.sh in "Build velox with substrait“ section as follow.
        name: Build velox with substrait
          command: |
           ./script/setup_ubuntu.sh
            mkdir -p .ccache
            export CCACHE_DIR=$(realpath .ccache)
            ccache -sz -M 5Gi
            source /opt/rh/gcc-toolset-9/enable
            make clean
            make release EXTRA_CMAKE_FLAGS="-DVELOX_ENABLE_SUBSTRAIT=ON" NUM_THREADS=16 MAX_HIGH_MEM_JOBS=8 MAX_LINK_JOBS=16
            ccache -s
           no_output_timeout: 1h

But I feel this way is weird.

File: .circleci/config.yml
Comment (majetideepak) [inline .circleci/config.yml:373]: can we invoke `setup-ubuntu.sh install_protobuf` ?

File: .circleci/config.yml
Comment (majetideepak) [inline .circleci/config.yml:414]: @kgpai Is Substrait support required to be part of the minimal build?

File: .circleci/config.yml
Comment (majetideepak) [inline .circleci/config.yml:425]: This leg is only supposed to be for building minimal Velox. Not for running tests.

**Comment by majetideepak** [comment]
@zhejiangxiaomai How long do Subtrait tests run? We can consider adding them to linux-parquet-hive-storage-adapters-build run. We can rename this to simply `linux-adapters`.

**Review by majetideepak — CHANGES_REQUESTED** [review]
@zhejiangxiaomai Sorry for the late review.

File: .circleci/config.yml
Comment (zhejiangxiaomai) [reply .circleci/config.yml:373]: No, The docker image of circleCI is centos, and there is some "apt" instruction at setup-ubuntu.sh.

File: .circleci/config.yml
Comment (zhejiangxiaomai) [reply .circleci/config.yml]: > @kgpai Is Substrait support required to be part of the minimal build?

Hi, @majetideepak  Substrait support is not required to be part of the minimal build.

File: .circleci/config.yml
Comment (zhejiangxiaomai) [reply .circleci/config.yml]: > This leg is only supposed to be for building minimal Velox. Not for running tests.

Nice comments, I move this part to linux-adapters.

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai Sorry for the late review.

You are welcome, thank you very much for your time.

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai How long do Subtrait tests run? We can consider adding them to linux-parquet-hive-storage-adapters-build run. We can rename this to simply `linux-adapters`.

Subtrait tests only take 2 minutes to run all tests. And it was added into linux-adapters now.

**Comment by zhejiangxiaomai** [comment]
Hi, @majetideepak  all comments were fixed. The major change is add install_protobuf in setup-adapters.sh. It make code more clear to review. And rename "linux-parquet-hive-storage-adapters-build" to "linux-adapters". Could you please review it again? And merge.

**Review by mbasmanova — COMMENTED** [review]
@zhejiangxiaomai Which CI job is running Substrait tests?

@majetideepak @kgpai Would you review the build script changes?

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai Which CI job is running Substrait tests?
> 
> @majetideepak @kgpai Would you review the build script changes?

Hi, @mbasmanova  Linux-adapters will run Substrait tests.  Because it compile with -DVELOX_ENABLE_SUBSTRAIT=ON. 
You can find them at [linux-adapters restult](https://circleci.com/api/v1.1/project/github/facebookincubator/velox/77060/output/111/0?file=true&allocation-id=631754c06ff245020970c2f7-0-build%2F60EC73FA).

And here is a snapshot of it.
image

**Comment by mbasmanova** [comment]
> Linux-adapters will run Substrait tests. 

Awesome. Please, update PR description to clarify.

`Add Substrait tests CircleCI job linux-adapters.`

**Comment by zhejiangxiaomai** [comment]
> > Linux-adapters will run Substrait tests.
> 
> Awesome. Please, update PR description to clarify.
> 
> `Add Substrait tests CircleCI job linux-adapters.`

Ok, thank you.

**Review by kgpai — APPROVED** [review]

**Comment by kgpai** [comment]
LG from my end; I also validated that protobuf version works on TA images. Please get approval from @majetideepak before I can start the merge process.

File: scripts/setup-adapters.sh
Comment (majetideepak) [inline scripts/setup-adapters.sh:51]: Is this required? Users usually run these scripts without reading. It might be undesired for Velox scripts to remove exisiting libraries.

File: scripts/setup-adapters.sh
Comment (majetideepak) [inline scripts/setup-adapters.sh:50]: Is this required to be part of this script? This should be part of the environmental setup?

Comment (zhejiangxiaomai) [reply scripts/setup-adapters.sh:50]: > Is this required to be part of this script? This should be part of the environmental setup?

It is necessary, we need install gcc-c++ to compile protobuf. we will get error when we compile protobuf without gcc-c++.

File: scripts/setup-adapters.sh
Comment (zhejiangxiaomai) [reply scripts/setup-adapters.sh]: > Is this required? Users usually run these scripts without reading. It might be undesired for Velox scripts to remove exisiting libraries.

It's not required, I remove this.

**Comment by zhejiangxiaomai** [comment]
Hi, @majetideepak I fix all comments, would you please review again? Thank you.

**Comment by mbasmanova** [comment]
@majetideepak Deepak, any chance you could take another look at this PR? I believe the comments have been addressed. I just want to make sure everything is correct before merging.

**Comment by zhejiangxiaomai** [comment]
> @zhejiangxiaomai I ran substrait-validator from https://github.com/substrait-io/substrait-validator on q6_first_stage.json file present in latest main and it shows a lot of errors. I wonder if some of the errors are being fixed in this PR. Would you run the validator and share the results? If there are still errors, let's prioritize fixing the JSON file to fix these.
> 
> substrait-validator substrait/tests/data/q6_first_stage.json --out-file /tmp/plan.html --mode ignore
> 
> Screen Shot 2022-09-13 at 10 17 22 AM

Ok, I will check this json.

**Comment by zhejiangxiaomai** [comment]
Hi, @mbasmanova I reproduced the problem  about "extension_uri_reference" by using substrait-validator. In fact, we do not use this field now, and I think this PR's aim is update protobuf and make CircleCI run substrait test. So can we merge this pr, And I will fix this errors in another PR later.

File: scripts/setup-adapters.sh
Comment (majetideepak) [inline scripts/setup-adapters.sh:50]: nit: clean up for another PR.

**Review by majetideepak — APPROVED** [review]
The build script changes look good. Thanks @zhejiangxiaomai

File: scripts/setup-adapters.sh
Comment (zhejiangxiaomai) [reply scripts/setup-adapters.sh:50]: > nit: clean up for another PR.

Thanks, will do it.

**Pull request closed without merging**
